src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1779,21 +1779,75 @@ _monodroid_counters_dump (const char *format, ...)
 	monoFunctions.counters_dump (XA_LOG_COUNTERS, counters);
 }
 
+static void
+load_assembly (MonoDomain *domain, JNIEnv *env, jstring_wrapper &assembly)
+{
+	timing_period total_time;
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING)))
+		total_time.mark_start ();
+
+	const char *assm_name = assembly.get_cstr ();
+	MonoAssemblyName *aname;
+
+	aname = monoFunctions.assembly_name_new (assm_name);
+
+	if (domain != monoFunctions.domain_get ()) {
+		MonoDomain *current = monoFunctions.domain_get ();
+		monoFunctions.domain_set (domain, FALSE);
+		monoFunctions.assembly_load_full (aname, NULL, NULL, 0);
+		monoFunctions.domain_set (current, FALSE);
+	} else {
+		monoFunctions.assembly_load_full (aname, NULL, NULL, 0);
+	}
+
+	monoFunctions.assembly_name_free (aname);
+
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING))) {
+		total_time.mark_end ();
+
+		timing_diff diff (total_time);
+		log_info (LOG_TIMING, "Assembly load: %s preloaded; elapsed: %lis:%lu::%lu", assm_name, diff.sec, diff.ms, diff.ns);
+	}
+}
+
+static void
+load_assemblies (MonoDomain *domain, JNIEnv *env, jstring_array_wrapper &assemblies)
+{
+	timing_period total_time;
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING)))
+		total_time.mark_start ();
+
+	/* skip element 0, as that's loaded in create_domain() */
+	for (size_t i = 1; i < assemblies.get_length (); ++i) {
+		jstring_wrapper &assembly = assemblies [i];
+		load_assembly (domain, env, assembly);
+	}
+
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING))) {
+		total_time.mark_end ();
+
+		timing_diff diff (total_time);
+		log_info (LOG_TIMING, "Finished loading assemblies: preloaded %u assemblies; wasted time: %lis:%lu::%lu", assemblies.get_length (), diff.sec, diff.ms, diff.ns);
+	}
+}
+
 static void
 monodroid_Mono_UnhandledException_internal (MonoException *ex)
 {
 	// Do nothing with it here, we let the exception naturally propagate on the managed side
 }
 
 static MonoDomain*
-create_and_initialize_domain (JNIEnv* env, jclass runtimeClass, jstring_array_wrapper &runtimeApks, jobjectArray assemblies, jobject loader, bool is_root_domain)
+create_and_initialize_domain (JNIEnv* env, jclass runtimeClass, jstring_array_wrapper &runtimeApks, jstring_array_wrapper &assemblies, jobject loader, bool is_root_domain)
```

Comment:
@grendello I take it this means we can't get rid of the list of assemblies being passed in from the Java side now?

Reply:
@dellis1972: that's *exactly* what this means

Reply:
alas